stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,300 @@
+  const Elf_Ehdr *elfHeader = nullptr;
+  const Elf_Shdr *symbolTable = nullptr;
+  const Elf_Shdr *stringTable = nullptr;
+
+public:
+  StaticBinaryELF(const char *filename) {
+    executablePathName(filename);
+    mmapExecutable(filename);
+  }
+
+  ~StaticBinaryELF() {
+    if (mapping != nullptr) {
+      munmap(const_cast<char *>(mapping), mapLength);
+    }
+  }
+
+  const char *getPathName() {
+    return fullPathName.c_str();
+  }
+
+  void *getSectionLoadAddress(const void *addr) {
+    if (elfHeader != nullptr) {
+      auto searchAddr = reinterpret_cast<Elf_Addr>(addr);
+      auto headers = reinterpret_cast<const Elf_Phdr *>(mapping +
+                                                        elfHeader->e_phoff);
+
+      for (size_t idx = 0; idx < elfHeader->e_phnum; idx++) {
+        auto header = &headers[idx];
+        if (header->p_type == PT_LOAD && searchAddr >= header->p_vaddr
+            && searchAddr <= (header->p_vaddr + header->p_memsz)) {
+          return reinterpret_cast<void *>(header->p_vaddr);
+        }
+      }
+    }
+    return nullptr;
+  }
+
+  // Lookup a function symbol by address
+  const Elf_Sym *findSymbol(const void *addr) {
+    if (symbolTable != nullptr) {
+      auto searchAddr = reinterpret_cast<Elf_Addr>(addr);
+      auto entries = symbolTable->sh_size / symbolTable->sh_entsize;
+      auto symbols = reinterpret_cast<const Elf_Sym *>(mapping +
+                                                       symbolTable->sh_offset);
+
+      for (decltype(entries) idx = 0; idx < entries; idx++) {
+        auto symbol = &symbols[idx];
+        if (ELF_ST_TYPE(symbol->st_info) == STT_FUNC
+            && searchAddr >= symbol->st_value
+            && searchAddr < (symbol->st_value + symbol->st_size)) {
+          return symbol;
+        }
+      }
+    }
+    return nullptr;
+  }
+
+  const char *symbolName(const Elf_Sym *symbol) {
+    if (stringTable == nullptr || symbol->st_name >= stringTable->sh_size) {
+      return nullptr;
+    }
+    return mapping + stringTable->sh_offset + symbol->st_name;
+  }
+
+private:
+  // If the binary is a symlink (eg /proc/self/exe) resolve to the canonical
+  // filename
+  void executablePathName(const char *filename) {
+    vector<char> fname(PATH_MAX);
+    ssize_t ret = readlink(filename, &fname[0], PATH_MAX-1);
+    if (ret == -1) {
+      fullPathName = filename;
+      return;
+    }
+    if (ret >= PATH_MAX) {
+      ret = PATH_MAX-1;
+    }
+    fname[ret] = '\0';
+    fullPathName = &fname[0];
+  }
+
+  // Parse the ELF binary using mmap to read it but keep the mapped region
+  // as small as possible and expand it as necessary
+  bool mmapExecutable(const char *filename) {
+    struct stat buf;
+
+    int fd = open(filename, O_RDONLY);
+    if (fd < 0) {
+      return false;
+    }
+
+    if (fstat(fd, &buf) != 0) {
+      close(fd);
+      return false;
+    }
+
+    void *map = mmap(nullptr, sizeof(Elf_Ehdr), PROT_READ, MAP_SHARED, fd, 0);
```

Comment:
Ive updated the code to parse ```/proc/self/maps``` with checks to see if the file has been deleted during execution. It also parses the ELF header/program headers that are in memory to avoid reading in that part of the file, but still ```mmap()s``` the section, symbol and string tables.